### PR TITLE
fix: stop the runner before restarting, restart on workspace config change

### LINKED ISSUE #5327: config file change doesn't trigger re-run in workspace
### Describe the bug

In the reproduction, changing neither `vitest.workspace.ts` nor `packages/*/vite.config.ts` doesn't trigger rerun. I thought this is a part of `forceRerunTriggers` https://vitest.dev/config/#forcereruntriggers, but it doesn't see the case.

### Reproduction

https://stackblitz.com/edit/vitest-dev-vitest-tkdqry?file=packages%2Flib2%2Fvite.config.ts

### System Info

```shell
System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.18.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm
    pnpm: 8.15.3 - /usr/local/bin/pnpm
  npmPackages:
    @vitest/ui: latest => 1.3.1 
    vite: latest => 5.1.4 
    vitest: latest => 1.3.1
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### LINKED ISSUE #6738: workspace + browser mode + watch leave server behind when root config changes
### Describe the bug

Running `vitest` worspace (in watch mode) with a project is using browser mode.

When a root config is changed `vitest` detects the change and try to restart.
But it doesn't first kill the previous browser runner, causing eventual failure:

```sh

Restarting due to config changes...

 DEV  v2.1.3 /home/unional/code/repobuddy/storybook
      [sb] Browser runner started by playwright at http://localhost:5173/

Port 5173 is in use, trying another one...

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Rejection ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Error: [vitest-api]: Timeout calling "createTesters"
 ❯ Object.onTimeoutError node_modules/.pnpm/@vitest+browser@2.1.3_@vitest+spy@2.1.3_playwright@1.48.1_typescript@5.6.3_vite@5.4.9_@types+node@22.7.5__vitest@2.1.3/node_modules/@vitest/browser/dist/index.js:577:17
 ❯ Timeout._onTimeout node_modules/.pnpm/@vitest+browser@2.1.3_@vitest+spy@2.1.3_playwright@1.48.1_typescript@5.6.3_vite@5.4.9_@types+node@22.7.5__vitest@2.1.3/node_modules/@vitest/browser/dist/index.js:240:41
 ❯ listOnTimeout node:internal/timers:573:17
 ❯ process.processTimers node:internal/timers:514:7

 ELIFECYCLE  Command failed with exit code 1.
```

### Reproduction

TBD, I have a repro but need to clean it up.

### System Info

```shell
System:
    OS: Linux 5.15 Ubuntu 22.04.3 LTS 22.04.3 LTS (Jammy Jellyfish)
    CPU: (16) x64 AMD Ryzen 7 5800X 8-Core Processor
    Memory: 28.36 GB / 31.31 GB
    Container: Yes
    Shell: 5.8.1 - /usr/bin/zsh
  Binaries:
    Node: 20.10.0 - /run/user/1000/fnm_multishells/621_1729231040440/bin/node
    npm: 10.2.3 - /run/user/1000/fnm_multishells/621_1729231040440/bin/npm
    pnpm: 9.12.1 - /run/user/1000/fnm_multishells/621_1729231040440/bin/pnpm
    bun: 1.0.22 - ~/.bun/bin/bun
  npmPackages:
    @vitejs/plugin-react: ^4.3.2 => 4.3.2 
    @vitest/browser: ^2.1.3 => 2.1.3 
    vite: ^5.4.9 => 5.4.9 
    vitest: ^2.1.3 => 2.1.3
```

### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by sheremet-va**
### Description

Fixes #6738
FIxes #5327

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: packages/vitest/src/node/cli/cli-api.ts [hunk 1]
```diff
@@ -73,8 +73,9 @@ export async function startVitest(
     stdinCleanup = registerConsoleShortcuts(ctx, stdin, stdout)
   }
 
-  ctx.onServerRestart((reason) => {
+  ctx.onServerRestart(async (reason) => {
     ctx.report('onServerRestart', reason)
+    await ctx.close()
```
Comment (AriPerkkio) [inline packages/vitest/src/node/cli/cli-api.ts:78]: I was looking at these specific lines earlier this week, and decided to move this logic into `core.ts` directly. We don't really need to attach this `onServerRestart` listener outside the ctx.

- https://github.com/AriPerkkio/vitest/commit/7e222e999314164e5a3d29c89897a1dc4bf6fdd9#diff-49ef635be88fe607c8682e81ab56b061ba9aafd5c94a5690a70b90a54604cd24L76-L79
- https://github.com/AriPerkkio/vitest/commit/7e222e999314164e5a3d29c89897a1dc4bf6fdd9#diff-c70ce04cc623628f74475cbde7bf56c29e5c3034b6fd5d26f737cb3944254463R160-R161

We could do this refactoring in this PR too.

Comment (sheremet-va) [reply packages/vitest/src/node/cli/cli-api.ts:78]: I don't really see any benefit in moving it to be honest 🤔 If we removed `ctx.onServerRestart` entirely - then maybe 🤷🏻

Comment (AriPerkkio) [reply packages/vitest/src/node/cli/cli-api.ts:78]: As I was looking all the places where we call reporters, this line looked surprising. Why are we attaching these listeners outside `core.ts`? Why don't we call `ctx.report('onServerRestart', reason)` directly inside `core.ts`? If there's a real reason for these lines to be outside it, here in `cli-api`, then sure let's leave them.

Comment (sheremet-va) [reply packages/vitest/src/node/cli/cli-api.ts]: Sure, I don't have enough arguments against it

**Review by hi-ogawa — COMMENTED** [review]
Looks good to me :+1: Tested locally with `pnpm -C examples/workspace test` and `pnpm -C test/browser test-fixtures`.
Can you add relevant tests?

File: packages/vitest/src/node/core.ts [hunk 2]
```diff
@@ -145,6 +150,8 @@ export class Vitest {
       const serverRestart = server.restart
       server.restart = async (...args) => {
         await Promise.all(this._onRestartListeners.map(fn => fn()))
+        this.report('onServerRestart', 'config')
```
Comment (AriPerkkio) [inline packages/vitest/src/node/core.ts:153]: Is the `'config'` reason here correct? On the line above it's not passed to `_onRestartListeners` calls like we do below on `server.watcher.on('change', ...` callback. 

```suggestion
        this.report('onServerRestart')
```

This affects here:

https://github.com/vitest-dev/vitest/blob/0d67f04b653f7a5f6971083d05367e5a2478a17d/packages/vitest/src/node/reporters/base.ts#L364-L374

Comment (sheremet-va) [reply packages/vitest/src/node/core.ts:153]: Yeah, I just didn't notice

**Comment by sheremet-va** [comment]
@hi-ogawa @AriPerkkio added the test. interested to know your opinion on the `useFS` helper

**Comment by AriPerkkio** [comment]
Looks handy and easy to read. Sometimes it's nicer to have static fixtures directory but in cases like this where they are not shared between multiple tests, re-creating fixtures on test run makes sense.

The `reruns tests when configs change` test case does a lot of file writing so it might be slow on CI. Looks like it did pass though.

**Comment by sheremet-va** [comment]
> The reruns tests when configs change test case does a lot of file writing so it might be slow on CI. Looks like it did pass though.

Maybe we can emit the event manually instead of waiting then 🤔

**Comment by sheremet-va** [comment]
It takes 1200-1300ms in CI, the same locally for me. Takes the longest on Windows - 1700ms, but changing to use `watcher.emit` didn't improve times significantly 🤷🏻

**Review by hi-ogawa — APPROVED** [review]
> interested to know your opinion on the `useFS` helper

It looks nice for small things and I like it :+1:
But for complicated setup, I prefer having files committed in the repo as it allows `vitest --root (some-fixture-dir)` to debug and iterate easily from cli.

Btw, currently root `fixtures/config-watching` nor any directories inside are not removed. Should we cleanup root entirely? (though I might even want a mode with env var where it leaves files without being deleted to debug later)

---

Just for future references, sharing a few projects I know doing something similar:
- Remix https://github.com/remix-run/remix/blob/1bd7f943c3f90c9e7e9ca5742324f1bb2622958c/integration/vite-dev-test.ts#L19
- Playwright https://github.com/microsoft/playwright/blob/50775698ae13642742f2a1e8983d1d686d7f192d/tests/playwright-test/test-grep.spec.ts#L20

From my experience, I wasn't fond of Remix's huge fixture as it was often hard to iterate and debug. But, looking at playwright's ones, they make use of a lot of small `runInlineTest` and it looks nice.

**Review by hi-ogawa — APPROVED** [review]
:+1:

**Comment by unional** [comment]
Hi, just want to share that I'm still seeing the port being used with 3.0.6.

**Pull request merged**